### PR TITLE
added a waitgroup to wait until all the pool connections are dialed

### PR DESCRIPTION
**Pull request by sredxny**
## Description

On connect to RPC do not loose control on when the initial connections is done (dial) as it might happens that gw cannot connect or reach the MDCB server but anyway attempts to login and pull resources (policies & apis). So, instead, wait until the dialing is succesful so gw do not start without apis loaded as described in the ticket

## Related Issue

https://tyktech.atlassian.net/browse/TT-12537

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

### **PR Type**
enhancement

___

### **Description**
- Added a `sync.WaitGroup` to ensure all RPC pool connections are dialed before proceeding with login and resource synchronization.
- Modified the `Connect` function to wait for all connections to be established, preventing premature login attempts.
- Improved the reliability of the gateway startup by ensuring that APIs are loaded only after successful connection dialing.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>rpc_client.go</strong><dd><code>Implement wait group for connection dialing synchronization</code></dd></summary>
<hr>

rpc/rpc_client.go

<li>Introduced a <code>sync.WaitGroup</code> to manage connection dialing.<br> <li> Modified the connection dialing process to wait for all connections to <br>be established before proceeding.<br> <li> Ensured that the <code>handleLogin</code> function is called only after all <br>connections are dialed.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6487/files#diff-3b88914c99bb9418e44e6389ce73579843562e8900730b380d7fff2e95c51033">+10/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by sredxny** [comment]
Sonnarcloud is complaining on lines that were not touched in this PR, so it needs to get adjusted but it will be done in a separate PR

File: rpc/rpc_client.go [hunk 1]
```diff
@@ -294,8 +298,10 @@ func Connect(connConfig Config, suppressRegister bool, dispatcherFuncs map[strin
 		conn.Write([]byte("proto2"))
 		conn.Write([]byte{byte(len(connID))})
 		conn.Write([]byte(connID))
+		connectionDialingWG.Done()
```
Comment (mativm02) [inline rpc/rpc_client.go:301]: let's add a `defer` to this line, and move it to the start of the function, just in case it doesn't reach here

**Comment by excieve** [comment]
@sredxny Thanks for this, but the other thing mentioned in the ticket — `handleLogin` is also asynchronous (it calls a goroutine inside). Does it address the case where connection exists but then policy/API sync happens before the RPC login?

**Pull request merged**